src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -622,7 +628,14 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 continue 'dfs;
             }
 
-            stack.extend(mir.predecessor_locations(l));
+            stack.extend(mir.predecessor_locations(l)
+                         .map(|predecessor| {
+                         // for each of l's predecessors, if 'l' is also a dominant, then that
+                         // implies that there is a back-edge from l to the predecessor and hence
+                         // we add the tuple (predecessor, true) or else (predecessor, false).
+                         (predecessor, l.dominates(predecessor, &self.dominators))
```

Comment:
The problem is that we want this boolean to be true if we traversed a back edge **at any point along the path**.  So something like this:

```rust
let is_back_edge = context.loc.dominates(predecessor, &self.dominators);
(predecessor,  l.1 || is_back_edge)
```